### PR TITLE
PLT-871: ThunkRecursions conservative mode

### PR DESCRIPTION
**Pull request by bezirg**
The current thunk recursions transformation has 2 problems:

1) does not preserve order of effects, so it may rearrange an error or a trace
2) does not respect the original strictness of some pure-strict bindings, turning them into lazy (not extremely bad)

This PR fixes (1) and (2) 

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Changelog fragments have been written (if appropriate)
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting master unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/plutus-ir/src/PlutusIR/Compiler.hs [hunk 1]
```diff
@@ -241,7 +249,7 @@ compileReadableToPlc (Program a v t) =
         >=> NonStrict.compileNonStrictBindings False
         >=> through check
         >=> (<$ logVerbose "  !!! thunkRecursions")
-        >=> (withBuiltinsInfo . fmap pure . flip ThunkRec.thunkRecursions)
+        >=> (\t' -> withBuiltinsInfo $ \semvar -> withCcOpts $ \opts -> pure $ ThunkRec.thunkRecursions semvar opts t')
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:252]: at this point it's easier to just write an inline `\t -> do ...`, I think?

File: plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs [hunk 2]
```diff
@@ -93,51 +94,74 @@ becomes
...
+    isStrictEffectful :: Binding tyname name uni fun a -> Bool
+    isStrictEffectful = \case
+        TermBind _ Strict _ rhs -> not $ isPure binfo vinfo rhs
+        _                       -> False
+
+    needsNonStrictify :: Binding tyname name uni fun a -> Bool
+    needsNonStrictify b =
+        if opts^.coPreserveEffectOrder
+        then isProblematic b || isStrictEffectful b
+        else isProblematic b
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs:149]: I'd like to see some discussion about why it's worth having two modes! Why not just do the conservative thing always? It doesn't look like a lot of savings: the advantage of non-conservative mode is that we keep the recursive binding strict, and so we avoid some overhead from that. But it's not clear to me how much difference that makes. And it only applies in a niche case where we have something that _is_ of function type (so doesn't get strictified anyway), but is impure.

So I'd be pretty inclined to just make the conservative mode the only mode.

**Review by michaelpj — COMMENTED** [review]
Change definitely makes sense. Various thoughts...

File: plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs [hunk 3]
```diff
@@ -93,51 +94,74 @@ becomes
...
     -- See Note [Thunking recursions]
-    let (toThunk, noThunk) = partition strictNonFunctionBinding bs
-        newBindings = fmap nonStrictify toThunk ++ noThunk
-        strictifiers = mapMaybe (mkStrictifyingBinding binfo vinfo) toThunk
-    in mkLet a Rec newBindings $ mkLet a NonRec strictifiers t
-thunkRecursionsStep _ _ t = t
+    let (toNonStrictify, rest) = NE.partition needsNonStrictify bs
+        -- MAYBE: use some prism/traversal to keep the original arrangement of the let group
+        -- this is not a semantic problem, but just a stylistic/debugging issue where the original
+        -- let-group will have reordered the (lazified) bindings
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs:133]: I am also not sure if i want to rely on a later pass

File: plutus-core/plutus-ir/test/TransformSpec.hs [hunk 4]
```diff
@@ -65,13 +64,28 @@ transform =
...
 thunkRecursions :: TestNested
 thunkRecursions =
-    testNested "thunkRecursions" $
-        map
-            (goldenPir (ThunkRec.thunkRecursions def) pTerm)
-            [ "listFold"
-            , "monoMap"
-            , "errorBinding"
-            ]
+    testNested "thunkRecursions"
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/plutus-ir/test/TransformSpec.hs:67]: I also thought of adding that, but I was hit by an unrelated trace missing bug

File: plutus-core/plutus-ir/src/PlutusIR/Compiler.hs (hunk 1 above)
Comment (bezirg) [reply plutus-core/plutus-ir/src/PlutusIR/Compiler.hs:252]: this whole thing sucks a bit. I pushed the monadreader compilationctx down to each pass and then I was super excited how the code looked great. But then half of the test cases broke, so i reverted it

File: plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs [hunk 5]
```diff
@@ -93,51 +93,69 @@ becomes
...
+        editedLet = mkLet a Rec $ fmap nonStrictifyB toNonStrictify ++ rest
+        -- we now have to consider for which of those non-strictified we are going to introduce strictifiers
+        -- to preserve the original strictness annotations, we have to introduce strictifiers *for all* those we previously non-strictified
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs:140]: The real reason is to preserve any potential effects, right?

**Review by michaelpj — APPROVED** [review]
lgtm

File: plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs (hunk 5 above)
Comment (bezirg) [reply plutus-core/plutus-ir/src/PlutusIR/Transform/ThunkRecursions.hs:140]: Indeed, this comment came from somewhere else. The comment made sense previously but not now. Fixing

**Pull request merged**